Add support for loading/serving non-OSM data

steinbro opened the pull request:
With these changes, zero or more CSV files can be included in the ingest image, which will be loaded into the database alongside the imported OSM data.

Here's a summary of the modifications I made:
1. Add a `non_osm_data` table to the database.
2.Add a clause to the SQL `soundscape_tile` function to include rows from `non_osm_data` in addition to the OSM data in the served map tiles.
3. Add an `--extradatadir` argument to ingest.py, which specifies an optional directory containing CSV files to be loaded into the `non_osm_data` table during database provisioning.
    1. Each CSV must have at least four columns: `latitude`, `longitude`, `feature_type`, and `feature_value`. All other columns will be saved/served as key/value pairs in the item's `properties`.
    2. Each row is assigned a negative value for `osm_id` to ensure it does not conflict with any real OSM objects. (This is apparently the ID assignment strategy OSM editors use for unsaved data.)
4. Add a (commented-out) `COPY` command to `Dockerfile.ingest` to document how to add CSV files to the image.

RDMurray commented on svcs/data/ingest.py:
```diff
@@ -160,6 +162,67 @@ async def provision_database_async(postgres_dsn, osm_dsn):
+
+async def import_non_osm_data_async(csv_dir, osm_dsn):
+    # The client expects OSM IDs for every point, but this is not OSM data.
+    # Assign large negative OSM IDs, which will not conflict with real values.
```
Do you know if the client requires these IDs to be globally unique? If so they could potentially clash with the places table which has negative osm_id values. It's because imposm uses negative ids for relations. https://imposm.org/docs/imposm3/latest/relations.html

steinbro replied:
I'm not positive about this. The [developer docs](https://github.com/soundscape-community/soundscape/blob/main/docs/services/data-plane-schema.md#soundscape-use-of-geojson) describes the `osm_ids` field as " array of OSM ids cooked to avoid the fact that OSM ids are not unique across nodes, ways, relations, and areas. For a feature that directly represents an OSM element, this will be an array of a single OSM id. (TODO: document standard cooking scheme)" Given that TODO, I guess we'll need to reference the source code ourselves.

steinbro replied:
Per your comment, I set the `osm_id` for `non_osm_data` to start at `1e17`.

RDMurray reviewed (commented):
Thanks for working on this.

steinbro commented on svcs/data/Dockerfile.ingest:
```diff
@@ -15,20 +15,37 @@ ENV IMPOSM_BINARY_RELEASE=0.11.1
 
 COPY requirements.txt requirements_kubernetes.txt ingest.py kubescape.py extracts.json tilefunc.sql $INGEST/
+# Add line(s) like the following to import data points from CSV files
+#COPY test_data.csv $EXTRADATA/
```
Excellent point, fixed.

RDMurray commented:
The imposm mapping specifies "use_single_id_space=true" which is an [undocumented feature](https://github.com/omniscale/imposm3/issues/240)

> - node = positive,
> -  way negative
> -  and relation negative < - 1e17.

From current database:
```
osm=# SELECT                                                                    
    MIN(osm_id) AS min_osm_id,                                                  
    MAX(osm_id) AS max_osm_id                                                   
FROM (                                                                          
    SELECT osm_id FROM osm_entrances                                            
    UNION ALL                                                                   
    SELECT osm_id FROM osm_places                                               
    UNION ALL                                                                   
    SELECT osm_id FROM osm_roads                                                
) AS combined;                                                                  
     min_osm_id      | max_osm_id                                               
---------------------+-------------                                             
 -100000000018160483 | 12249301939                                              
(1 row)                                                                         
```

So maybe using IDs > 1E17 would be a good option?

RDMurray commented:
@steinbro  Are you happy to merge this now?

RDMurray commented on svcs/data/ingest.py:
```diff
@@ -198,6 +262,10 @@ def execute_kube_updatemodel_provision_and_import(config, updated):
                 provision_database(dsn_init, dsn)
+                if config.extradatadir:
+                    logger.info('Importing non-OSM data: START')
+                    import_non_osm_data(config.extradatadir, dsn)
```
This is only exicuted the first time round the loop, when the db status is None. it will also modify the current public schema which will be overwritten when the schemas are rotated by imposm. I think it should be below, just after the call to import_rotate().

steinbro replied:
Nice catch; I agree. Fixed.

steinbro commented:
@RDMurray We might as well wait to merge until #144 is ready. We'll probably want to tweak the columns in the CSV conversion here depending on what tag we settle on for classifying the non-OSM data points in the app.

JJGatchalian commented:
After the Non-OSM data code gets merged, is the regular OSM data getting updated anytime soon? I don't think it got changed since Soundscape Community version 1.20 back in October. I have added and deleted some places to my city in OSM, but these changes currently aren't reflected in Soundscape. Or are you timing it so that a major version change like the upcoming 1.30 gets released and then the database gets updated?

steinbro commented:
> After the Non-OSM data code gets merged, is the regular OSM data getting updated anytime soon? I don't think it got changed since Soundscape Community version 1.20 back in October. I have added and deleted some places to my city in OSM, but these changes currently aren't reflected in Soundscape. Or are you timing it so that a major version change like the upcoming 1.30 gets released and then the database gets updated?

The OSM data should be updating every week or two, unless the pipeline is getting stuck somewhere. @RDMurray can you check on the last DB refresh time, or check for errors in the ingest service if it hasn't been succeeding for a while?

RDMurray commented:
Oops, yes you are right it last successfully updated in October. The wget command got a 404 in november, and the unhandled exception terminated the ingest service. I restarted it for now and it is downloading, but I think the ingest service needs an overhall to make it more robust, add monitoring and strip out all the cruft left over from when microsoft used to manage several instances of Postgres in Kubernetes.

steinbro commented:
> I think the ingest service needs an overhall to make it more robust, add monitoring and strip out all the cruft left over from when microsoft used to manage several instances of Postgres in Kubernetes.

I agree. Wanna create one or more issues explaining the enhancements you think we should tackle? There could be a good student project in there, or at least some technical contribution opportunities that don't require Xcode or iOS development knowledge.

JJGatchalian commented:
Just curious. How long does the ingest process take? You said you started it, and the comment was posted arount 2:00 A.M. my time. When are we going to see the changes?

steinbro commented:
> Just curious. How long does the ingest process take? You said you started it, and the comment was posted arount 2:00 A.M. my time. When are we going to see the changes?

I think it's somewhere under 12 hours, but note that data you've requested before would be cached on your device. You can force the data in the app to be refreshed by clearing the cache under Settings > Troubleshooting.

RDMurray commented:
It's still going, but I don't know exactly how long it will take. It's a huge amount of data to load into the database and of course OSM is growing all the time.

JJGatchalian commented:
Thanks, I always clear the stored map data every day that I use Soundscape, first thing in the morning. I was guilty of deleting my own markers when using the app in Spanish so are we going to remove that prompt like the version released in June? Even assistive technology instructors make mistakes.

RDMurray commented:
Good point @JJGatchalian  I'll merge just that part of the original PR and upload another build before release.

steinbro commented:
I think we can go ahead and merge this too. When we want to actually include the NaviLens data in our backend, we'll need to run the CSV conversion script included in this commit and put the output in the extra_data directory on the server, but we can coordinate that separately from the merge.
